### DISCUSSION TITLE
preventing rebuilds

### DISCUSSION THREAD

**Discussion by qaristote** [nix-community/stylix#521] [Q&A] [answered]
Often when I update my system the theme generated by stylix gets rebuilt as well. This is expected behavior, I assume it happens either because GHC was updated or because the palette-generator code changed (I update stylix frequently). 

This is a bit inconvenient, as palette-generator takes a little while to generate a new theme, which is of course useless because it always happen to be the same as the previous one. So I was wondering if there were a way to tell NixOS to not rebuild the stylix-generated theme, or at least the palette-generator package (I assume the theme's only dependency is palette-generator). Is there a standard way to do this?

**Comment by trueNAHO** [comment]
> Often when I update my system the theme generated by stylix gets rebuilt as well. This is expected behavior, I assume it happens either because GHC was updated or because the palette-generator code changed (I update stylix frequently).
>
> [...]
>
> So I was wondering if there were a way to tell NixOS to not rebuild the stylix-generated theme, or at least the palette-generator package (I assume the theme's only dependency is palette-generator). Is there a standard way to do this?

AFAIK, the following snippet is responsible for caching the palette:

https://github.com/danth/stylix/blob/c5f8f06543b70248a076f888177c7362a24d5dcc/stylix/palette.nix#L169-L193

TBH, I am not entirely sure how `palette-generator` is actually cached. @danth might know more about this.

> This is a bit inconvenient, as palette-generator takes a little while to generate a new theme

Based on the discussion from https://github.com/danth/stylix/pull/477 it might be necessary to rewrite the palette generator in the future. Later, it might be possible to improve the palette generator performance.

**Reply by danth** [reply]
It seems that at some point this code was changed to generate the file through `config.lib.stylix.colors` rather than exporting [`generatedJSON`](https://github.com/danth/stylix/blob/c5f8f06543b70248a076f888177c7362a24d5dcc/stylix/palette.nix#L8) directly. That will have broken the caching behaviour. The intention was to make the `generatedJSON` derivation part of the final system so it can't be garbage collected between evaluations.

**Comment by qaristote** [comment, marked as answer]
I ended up setting `stylix.base16scheme` to the value I found in `/etc/stylix/generated.json`.
